contracts/hydro/src/vote.rs

```diff
@@ -0,0 +1,312 @@
+                    StdError::generic_err(
+                        "Failed to convert Decimal to SignedDecimal for validator shares",
+                    )
+                })?;
+
+            removed_votes.insert(lock_id, existing_vote.clone());
+
+            // Always remove vote from Vote Map.
+            // We cannot rely on it being overriden by the new vote (if any), as we don't know if it won't be skipped
+            VOTE_MAP.remove(storage, ((round_id, tranche_id), sender.clone(), lock_id));
+
+            // Remove voting round allowed info
+            VOTING_ALLOWED_ROUND.remove(storage, (tranche_id, lock_id));
+        }
+    }
+
+    Ok(ProcessUnvotesResult {
+        power_changes,
+        removed_votes,
+        locks_to_skip,
+    })
+}
+
+#[derive(Debug)]
+pub struct ProcessVotesResult {
+    pub power_changes: HashMap<u64, ProposalPowerUpdate>, // prop_id -> ProposalPowerUpdate
+    pub voted_proposals: Vec<u64>,
+    pub locks_voted: Vec<u64>,
+    pub locks_skipped: Vec<u64>,
+}
+
+// Struct to encapsulate the context for vote processing
+pub struct VoteProcessingContext<'a> {
+    pub env: &'a Env,
+    pub constants: &'a Constants,
+    pub sender: &'a Addr,
+    pub round_id: u64,
+    pub tranche_id: u64,
+}
+
+// Process votes (after unvote was processed)
+// It receives an argument as ProposalToLockups, which is a struct that contains a proposal_id and a list of lock_ids.
+// It also receives a list of lock_entries, which is a mapping of lock_id -> lock_entry.
+// It also receives a list of locks_to_skip, which is a list of lock_ids that should be skipped
+//  (as it was determined during process_unvotes that ).
+pub fn process_votes(
+    deps: &mut DepsMut<NeutronQuery>,
+    context: VoteProcessingContext,
+    proposals_votes: &[ProposalToLockups],
+    lock_entries: &LockEntries,
+    locks_to_skip: HashSet<u64>,
+) -> Result<ProcessVotesResult, ContractError> {
+    let round_end = compute_round_end(context.constants, context.round_id)?;
+    let lock_epoch_length = context.constants.lock_epoch_length;
+
+    let mut locks_voted = vec![];
+    let mut locks_skipped = vec![];
+    let mut voted_proposals = vec![];
+    let mut power_changes: HashMap<u64, ProposalPowerUpdate> = HashMap::new();
+
+    for proposal_to_lockups in proposals_votes {
+        let proposal_id = proposal_to_lockups.proposal_id;
+        let proposal = PROPOSAL_MAP.load(
+            deps.storage,
+            (context.round_id, context.tranche_id, proposal_id),
+        )?;
+
+        for &lock_id in &proposal_to_lockups.lock_ids {
+            // When instructed to skip the lock_id by process_unvotes, skip it and record as skipped
+            if locks_to_skip.contains(&lock_id) {
+                locks_skipped.push(lock_id);
+                continue;
+            }
+
+            let lock_entry = &lock_entries[&lock_id];
+
+            // If user didn't yet vote with the given lock in the given round and tranche, check
```

Comment:
The comment seems outdated now. We don't have an explicit check anymore for whether the user has voted in the given round/tranche; if they already voted in this round, then implicitly their old VOTING_ALLOWED_ROUND was erased during process_unvotes